### PR TITLE
Pennylane conversion

### PR DESCRIPTION
**Pull request by le-big-mac**
Adds scaling, corrects post-selection, and allows sympy symbols to be included in the translation.

### CODE REVIEW

File: discopy/quantum/pennylane.py [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+    OpType.CU1: lambda a, wires: qml.ctrl(qml.U1(a, wires=wires[1]),
+                                          control=wires[0]),
+    OpType.SWAP: qml.SWAP,
+    OpType.noop: qml.Identity,
+}
+
+
+def tk_op_to_pennylane(tk_op, str_map):
+    """
+    Extracts the operation, parameters and wires from
```
Comment (y-richie-y) [inline discopy/quantum/pennylane.py:38]: Extract the operation

File: discopy/quantum/pennylane.py [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+    a pytket `Op`, and return the corresponding PennyLane operation.
+
+    Parameters
+    -----------
```
Comment (y-richie-y) [inline discopy/quantum/pennylane.py:42]: Too many `-`

File: discopy/quantum/pennylane.py [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+        tk_op : pytket.circuit.Op
+            The pytket `Op` to convert.
+        str_map : dict[str, sympy.core.symbol]
+            A mapping from strings to SymPy symbols (necessary as
+            `circ.to_tk()` does not copy symbol references).
+
+    Returns
+    --------
```
Comment (y-richie-y) [inline discopy/quantum/pennylane.py:50]: Too many `-`

File: discopy/quantum/pennylane.py [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+            param = param.subs(sym_subs)
+        else:
+            param = torch.tensor([param])
+
+        remapped_params.append(param)
+
+    return OP_MAP[tk_op.op.type], remapped_params, wires
+
+
+def extract_ops_from_tk(tk_circ, str_map):
```
Comment (y-richie-y) [inline discopy/quantum/pennylane.py:75]: Extract

File: discopy/quantum/pennylane.py [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+    """
+    Extracts the operations, and corresponding parameters and wires,
+    from a pytket Circuit. Returns these as lists to use in
+    constructing PennyLane circuit.
+
+    Parameters
+    -----------
```
Comment (y-richie-y) [inline discopy/quantum/pennylane.py:82]: Too many `-`

File: discopy/quantum/pennylane.py [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+        tk_circ : :class:`discopy.quantum.tk.Circuit`
+            The pytket circuit to extract the operations from.
+        str_map : dict[str, sympy.core.symbol]
+            A mapping from strings to SymPy symbols.
+
+    Returns
+    --------
```
Comment (y-richie-y) [inline discopy/quantum/pennylane.py:89]: Too many `-`

File: discopy/quantum/pennylane.py [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+        self.probs = probabilities
+        self._post_selection = post_selection
+        self.scale = scale
+        self.n_qubits = n_qubits
+        self.device = device
+        self._valid_states = self.get_valid_states()
+
+    @property
+    def post_selection(self):
+        """
```
Comment (y-richie-y) [inline discopy/quantum/pennylane.py:195]: These can be single line docstrings

**Review by y-richie-y — COMMENTED** [review]
Some stylistic comments

File: discopy/quantum/pennylane.py [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+            params = [torch.cat(p) if len(p) > 0 else p
+                      for p in self.params]
+
+        wires = qml.draw(self.make_circuit())(params).split("\n")
+        for k, v in self._post_selection.items():
+            wires[k] = wires[k].split("┤")[0] + "┤" + str(v) + ">"
+
+        print("\n".join(wires))
+
+    def get_valid_states(self):
```
Comment (y-richie-y) [inline discopy/quantum/pennylane.py:252]: No need to have both get_valid_states and valid_states as a property.

File: discopy/quantum/pennylane.py [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+        Returns
+        --------
+            bool
+                Whether the circuit parameters contain SymPy symbols.
+        """
+        for expr_list in self.params:
+            if any(isinstance(expr, sympy.Expr) for
+                   expr in expr_list):
+                return True
+        return False
```
Comment (ianyfan) [inline discopy/quantum/pennylane.py:226]: `return any(isinstance(expr, sympy.Expr) for expr_list in self.params for expr in expr_list)` may be more concise

File: discopy/quantum/pennylane.py [hunk 10]
```diff
@@ -0,0 +1,379 @@
...
+        print("\n".join(wires))
+
+    def get_valid_states(self):
+        """
+        Determines which of the output states of the circuit are
+        compatible with the post-selections.
+
+        Returns
+        --------
+            list[int]
```
Comment (ianyfan) [inline discopy/quantum/pennylane.py:259]: Unindent

**Review by y-richie-y — APPROVED** [review]

**Comment by toumix** [comment]
Great work, congrats @le-big-mac on your first big DisCoPy PR 👏 👏 👏 
Do you think you could document the new features in a short Jupyter notebook?
That would also be the best way to advertise it (e.g. in the docs, on Twitter, etc.)

**Pull request merged**